Fix #2876: Create domain controller for feedback reporting

jcqli opened the pull request:
<!--
  - Thanks for submitting code to Oppia! Please fill out the following as part of
  - your pull request so we can review your code more easily.
  -->

## Explanation
Fix #2876
- Creates the domain controller that collects the feedback report information and either stores it on local disk or pushes it to remote storage. Defines functions to store, fetch, and remove reports from local storage, and send reports to the remote service.
- Adds a new data class object to represent the entry point.
- Add a language code extension to switch between app language, audio language, and a generic language code. No tests were added for the extensions as it seemed trivial to check hard-coded string mappings.
- Add an annotation class for the current report schema version and tests to validate the schema version and expected resort data.
- Add a 'platform' field to GaeFeedbackReport so backend schema validators can properly generalize to both report types.

Feature tracker: #53
[Design doc](https://docs.google.com/document/d/1_RTzad6u4m-ZFn0VTWNddp4bc0fCYrMk-ABIVYVJ7gU/edit?usp=sharing)

## Checklist
<!-- Please tick the relevant boxes by putting an "x" in them. -->
- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary auto-generated code from Android Studio.
- [ ] The PR is made from a branch that's **not** called "develop".
- [x] The PR is made from a branch that is up-to-date with "develop".
- [x] The PR's branch is based on "develop" and not on any other branch.
- [x] The PR is **assigned** to an appropriate reviewer in both the **Assignees** and the **Reviewers** sections.

jcqli committed:
Create protos for feedback report

jcqli committed:
Create protos for feedback report

jcqli committed:
Add bool to app startup state for crash

jcqli committed:
Add bool to app startup state for crash

jcqli committed:
Merge remote-tracking branch 'upstream/develop' into create-feedback-reporting-data-objects

jcqli committed:
Define data objects and service for sending feebdack reports

jcqli committed:
Define data objects and service for sending feebdack reports

jcqli committed:
Fix typo

jcqli committed:
Fix typo

jcqli committed:
Include unspecified for enum types

jcqli committed:
Include unspecified for enum types

jcqli committed:
Add protos to model bazel

jcqli committed:
Add protos to model bazel

jcqli committed:
Merge remote-tracking branch 'upstream/develop' into create-feedback-reporting-model-layer

jcqli committed:
Fix typos

jcqli committed:
Fix typos

jcqli committed:
Merge branch 'create-feedback-reporting-data-objects' into create-data-layer-componets-feedback-reporting

jcqli committed:
Create data class objets for feedback reporting

jcqli committed:
Create data class objets for feedback reporting

jcqli committed:
Remove topic database from feedback reporting proto

jcqli committed:
Remove topic database from feedback reporting proto

jcqli committed:
Merge branch 'create-feedback-reporting-data-objects' into create-data-layer-componets-feedback-reporting

jcqli committed:
Finish adding all data class objects

jcqli committed:
Finish adding all data class objects

jcqli committed:
Fix typo

jcqli committed:
Fix typo

jcqli committed:
Clarify comment

jcqli committed:
Clarify comment

jcqli committed:
Apply review comments

jcqli committed:
Apply review comments

jcqli committed:
Clarify proto enums

jcqli committed:
Clarify proto enums

jcqli committed:
Merge branch 'create-feedback-reporting-data-objects' into create-data-layer-componets-feedback-reporting

jcqli committed:
Update data class models after, add API key to NetworkModule

jcqli committed:
Update data class models after, add API key to NetworkModule

jcqli committed:
Create service mock and fake JSON data for testing

jcqli committed:
Create service mock and fake JSON data for testing

jcqli committed:
Add newline at eof

jcqli committed:
Add newline at eof

jcqli committed:
Add newline at eof

jcqli committed:
Add newline at eof

jcqli committed:
Write service tests

jcqli committed:
Write service tests

jcqli committed:
Merge remote-tracking branch 'upstream/develop' into create-data-layer-componets-feedback-reporting

jcqli committed:
Update timestamp data type to be double

jcqli committed:
Update timestamp data type to be double

jcqli committed:
Linter

jcqli committed:
Linter

jcqli committed:
Undo changes to cache store test

jcqli committed:
Undo changes to cache store test

jcqli committed:
Fix linter errors

jcqli committed:
Fix linter errors

jcqli committed:
Remove comma at end of data class Json items

jcqli committed:
Adjust service defintion

jcqli committed:
Merge branch 'create-data-layer-componets-feedback-reporting' of github.com:jcqli/oppia-android into create-data-layer-componets-feedback-reporting

jcqli committed:
Merge remote-tracking branch 'jcqli/create-data-layer-componets-feedback-reporting' into create-data-layer-for-feedback-reporting

jcqli committed:
Create oneof for each issue category

jcqli committed:
Address review comments

jcqli committed:
Merge remote-tracking branch 'jcqli/create-feedback-reporting-data-objects' into create-data-layer-for-feedback-reporting

jcqli committed:
Adjust test to execute and check for success

jcqli committed:
Update data classes to match protobufs

jcqli committed:
Remove unused imports

jcqli committed:
Clarify field names / data

jcqli committed:
Merge remote-tracking branch 'upstream/develop' into create-feedback-reporting-data-objects

jcqli committed:
Address review comments

jcqli committed:
Clarify locale

jcqli committed:
Apply review comments

jcqli committed:
Rename to AppContext

jcqli committed:
Add typo to issue

jcqli committed:
Merge remote-tracking branch 'origin/develop' into create-data-layer-for-feedback-reporting

jcqli committed:
Merge remote-tracking branch 'jcqli/create-feedback-reporting-data-objects' into create-data-layer-for-feedback-reporting

jcqli committed:
Update GAE models to match protos, fix names, add logs to app context

jcqli committed:
Fix linter error

jcqli committed:
Clarify country locale

jcqli committed:
Update service test to use JSON to make GAE object

jcqli committed:
Merge remote-tracking branch 'jcqli/create-feedback-reporting-data-objects' into create-data-layer-for-feedback-reporting

jcqli committed:
Stub controller

jcqli committed:
Remove controller from this PR

jcqli committed:
Start management controller for feedback reporting

jcqli committed:
Apply review comments

jcqli committed:
Add api key and authorization class

jcqli committed:
Remove api key from network module

jcqli committed:
Add KDoc comments to explain GAE model fields

jcqli committed:
Fix typo in comment

jcqli committed:
Merge remote-tracking branch 'origin/develop' into create-data-layer-for-feedback-reporting

jcqli committed:
Fix typo

jcqli committed:
Update URL path for service

jcqli committed:
Add functions to report management controller that fetches / saves reports in local store

jcqli committed:
Provide implementation for feedback reporting service

jcqli committed:
Apply review comments

jcqli committed:
Merge branch 'create-data-layer-for-feedback-reporting' into create-domain-controller-for-feedback-reporting

jcqli committed:
Fix typo

jcqli committed:
Add service call to controller

jcqli committed:
Remove api key from moshi object to use in interceptor for request header

jcqli committed:
Adjust comment

jcqli committed:
Change JSON field name to match backend

jcqli committed:
Implement cases for different issues for  user supplied info

jcqli committed:
Add context info to controller

jcqli committed:
Get event logs

jcqli committed:
Fix error message

jcqli committed:
Add headers using network interceptor

jcqli committed:
Create event logs

jcqli committed:
Merge remote-tracking branch 'origin/develop' into create-domain-controller-for-feedback-reporting

jcqli committed:
Resolve conflicts

jcqli committed:
Change topic_progress to completed_exploration_ids

jcqli committed:
Fix comment

jcqli committed:
Merge remote-tracking branch 'origin/develop' into create-data-layer-for-feedback-reporting

jcqli committed:
Add entry point as a data class

jcqli committed:
Add entry point in controller

jcqli committed:
Start tests

jcqli committed:
Merge branch 'develop' into create-data-layer-for-feedback-reporting

jcqli committed:
Merge branch 'create-data-layer-for-feedback-reporting' into create-domain-controller-for-feedback-reporting

jcqli committed:
Fix linter errors

jcqli committed:
Add comments to clarify controller

jcqli committed:
Configure bazel?'

jcqli committed:
Bazel build?

BenHenning commented on domain/src/test/java/org/oppia/android/domain/feedbackreporting/FeedbackReportManagementControllerTest.kt:
```diff
@@ -0,0 +1,289 @@
+    @Provides
+    @Singleton
+    fun provideFeedbackReportingService(
+      @OppiaRetrofit retrofit: Retrofit
```
I think your compiler error is due to there not being any provider for this value (e.g. nothing is providing an OppiaRetrofit-qualified Retrofit object).

jcqli committed:
Clarify country code for locales

jcqli committed:
Specify language code'

jcqli committed:
Fix names to match JSON

jcqli committed:
Mock retrofit and service in test

jcqli committed:
Move mock services to testing module

jcqli committed:
Update data fields to match backend

jcqli committed:
Read logcat logs from file

jcqli committed:
Test builds

jcqli committed:
Remove completed explorations list

jcqli committed:
Make fake file for logcat

jcqli committed:
Init mocks in test

jcqli committed:
Update codeowners and fix linter errors

jcqli committed:
Undo bazel changes

jcqli committed:
Move mock services and api utils to testing module

jcqli committed:
Remove extra newline

jcqli committed:
Change field names to match backend

jcqli committed:
Use class const for language codes

jcqli committed:
Explain logcat file helper in a comment

jcqli committed:
Fix bazel build

jcqli committed:
Merge remote-tracking branch 'origin/develop' into create-domain-controller-for-feedback-reporting

jcqli committed:
Add moshi for annotations in bazel?

jcqli committed:
Remove moshi from bazel

jcqli committed:
Change audio language code to audio language

jcqli committed:
Update data classes to match backend declarations

jcqli committed:
Fix linter errors

jcqli committed:
Use Callback for request

jcqli committed:
Fix linter errors

jcqli committed:
Create language code extensions

jcqli committed:
Linter fixes

jcqli committed:
Use Call in service, change timestamp to int for seconds

jcqli committed:
Fix JSON test errors

jcqli committed:
Cleanup comments, code

jcqli committed:
Fix codeowners for JSON in data module

jcqli committed:
Move JSON to data module

jcqli committed:
Merge remote-tracking branch 'origin/develop' into create-domain-controller-for-feedback-reporting

BenHenning committed:
Enable Moshi support in Bazel.

This includes some reorganizing of the data module, but it introduces
support for generating GAE models through Moshi. Corresponding mock
services & tests are still failing, but //data & //:oppia build.

BenHenning committed:
Make mock services build for testing module.

BenHenning committed:
Add working data tests to Bazel build.

This also fixes a regression from an earlier commit that led to the
Moshi objects not actually being generated correctly, so the data
service tests are good proof that Moshi is being used correctly.

BenHenning committed:
Remove TODOs for #1675.

BenHenning committed:
Lint fixes.

BenHenning committed:
Merge branch 'develop' into enable-data-module-in-bazel

Conflicts:
	.bazelrc

BenHenning committed:
Bazel lint fix.

Avoid relying on native java_library.

BenHenning committed:
Fix broken Gradle build.

This includes explicitly specifying okhttp version since Gradle was
previously randomly picking one.

BenHenning committed:
Add debug details for affected tests script.

BenHenning committed:
Lint fixes.

BenHenning committed:
Attempt to force Bazel cache to clear.

This is an attempt to work around a theorized OOM happening when trying
to compute the sibling build graph for changes build files.

jcqli committed:
Merge branch 'develop' into create-domain-controller-for-feedback-reporting

jcqli committed:
Merge branch 'enable-data-module-in-bazel' into create-domain-controller-for-feedback-reporting

jcqli committed:
Create module for report schema version

jcqli committed:
Add tests for schema validator

jcqli committed:
Add FeedbackReportingModule to the application component

jcqli committed:
Adjust injection for schema version, tests build and run

jcqli committed:
Add test to check fields of high-level feedback report

jcqli committed:
Add field checks for all data classes

jcqli committed:
Use helper to verify data classes

jcqli committed:
Add entry point to model bazel, remove items from domain bazel

jcqli committed:
Merge remote-tracking branch 'origin/develop' into create-domain-controller-for-feedback-reporting

jcqli committed:
Add bazel dependencies

jcqli committed:
Make when statement exhaustive with no-op blocks

jcqli committed:
Remove imports and udpate JSON field names

jcqli committed:
Nit fixes

jcqli committed:
Merge branch 'develop' into create-domain-controller-for-feedback-reporting

jcqli committed:
Nit fixes

jcqli committed:
Update POST url

jcqli committed:
Clean up test

jcqli committed:
Merge remote-tracking branch 'origin/develop' into create-domain-controller-for-feedback-reporting

jcqli committed:
Remove duplicate dependency naming

jcqli committed:
Add MockWebServer to controller test

jcqli committed:
Fix linter errors

jcqli committed:
Specify language code to string, add profile proto to utility bazel

jcqli committed:
Fix import

jcqli committed:
Remove controller change from this branch

jcqli committed:
Revert "Remove controller change from this branch"

This reverts commit 92971344e86249d304950898b45b63bbc6af70fe.

jcqli committed:
Merge remote-tracking branch 'origin/develop' into create-domain-controller-for-feedback-reporting

jcqli committed:
Update json names

jcqli committed:
Use mock web server in controller test, update JSON to match backend

jcqli committed:
Use mock web server for all tests

jcqli committed:
Use retrofit for service

jcqli committed:
Add prod files to test implementation builds?

jcqli committed:
Fix dagger dependencies?

jcqli committed:
Fix linter issues

jcqli committed:
Merge remote-tracking branch 'origin/develop' into create-domain-controller-for-feedback-reporting

jcqli committed:
Merge remote-tracking branch 'origin/develop' into create-domain-controller-for-feedback-reporting

jcqli committed:
Update to match backend domin / storage

BenHenning commented:
@jcqli spent some time looking into the failures. The test via Bazel was experiencing a runtime failure due to not having access to the data test assets (for API mocks), and the test via Gradle was experiencing a build error due to the data module not running the Dagger compiler. The following gist should fix both: https://gist.github.com/BenHenning/da7d1eeb28360619c56ecc91367d4197/revisions. Confirming all non-app tests pass locally for me now on this branch via Bazel, and domain tests pass via Gradle.

Also, note that your test ``testController_submitFeedbackReport_withNetwork_sendsNetworkRequest_isSuccessful`` takes almost 10 seconds to run for me--that seems quite long. It's also not Robolectric arrangement since the first test took the expected ~8 seconds to start up. It may be worth looking into why this test in particular is taking so long.

jcqli committed:
Merge remote-tracking branch 'origin/develop' into create-domain-controller-for-feedback-reporting

jcqli committed:
Merge remote-tracking branch 'origin/develop' into create-domain-controller-for-feedback-reporting

jcqli committed:
Fix dependencies, add local annotations to define mock web server and retrofit instances

jcqli committed:
Locally inject network dependencies to properly set up mock server

jcqli commented:
> Also, note that your test `testController_submitFeedbackReport_withNetwork_sendsNetworkRequest_isSuccessful` takes almost 10 seconds to run for me--that seems quite long. It's also not Robolectric arrangement since the first test took the expected ~8 seconds to start up. It may be worth looking into why this test in particular is taking so long.

@BenHenning  I did a bit of guess and check and I think it's because of the timeout specified -- when I shorten the timeout amount the test runs faster, and when I increase it slows down accordingly. Might warrant some more research, though since I'm not sure if this is sound reasoning.

jcqli committed:
Fix controller test to use mock web server

jcqli committed:
Fix bazel build errors

jcqli committed:
Fix controller bazel tests

jcqli committed:
Fix bazel for data tests

jcqli committed:
Fix newline, remove generated changes

jcqli committed:
Fix format errors in bazel build files

jcqli committed:
Merge remote-tracking branch 'origin/develop' into create-domain-controller-for-feedback-reporting

jcqli committed:
Maybe fix linter errors

jcqli commented:
Hm I can't reproduce the bazel build error locally but it doesn't seem like a flake. I tried `bazel clean` and then rebuilding it but it looks good on mine, @BenHenning could you try reproducing this?

EDIT: I guess this might be a maven / network issue?

BenHenning commented:
Sorry, been really busy with non-work items this week. Will need to take a look later in the week.

BenHenning commented:
> > Also, note that your test `testController_submitFeedbackReport_withNetwork_sendsNetworkRequest_isSuccessful` takes almost 10 seconds to run for me--that seems quite long. It's also not Robolectric arrangement since the first test took the expected ~8 seconds to start up. It may be worth looking into why this test in particular is taking so long.
> 
> @BenHenning I did a bit of guess and check and I think it's because of the timeout specified -- when I shorten the timeout amount the test runs faster, and when I increase it slows down accordingly. Might warrant some more research, though since I'm not sure if this is sound reasoning.

I wouldn't expect the timeout to make a difference--that is a bit odd.



> Hm I can't reproduce the bazel build error locally but it doesn't seem like a flake. I tried `bazel clean` and then rebuilding it but it looks good on mine, @BenHenning could you try reproducing this?
> 
> EDIT: I guess this might be a maven / network issue?

@jcqli I think you're running into #3168. Syncing to the latest develop should fix CI for you.

BenHenning commented:
@jcqli does this PR have unrelated changes in it from earlier PRs? Some of the data module changes seem extra--could you sync to the latest develop and send back?

jcqli committed:
Merge remote-tracking branch 'origin/develop' into create-domain-controller-for-feedback-reporting

jcqli commented:
> I wouldn't expect the timeout to make a difference--that is a bit odd.

Now I can't reproduce this behavior, the test `testController_submitFeedbackReport_withNetwork_successfullySendsRequestToServer` takes about 4-5 seconds regardless of timeout on Gradle, is there a way to see individual run-times with Bazel?

> @jcqli does this PR have unrelated changes in it from earlier PRs? Some of the data module changes seem extra--could you sync to the latest develop and send back?

The data module Bazel file changes are largely from the earlier gist that resolved the build errors, but the GAE models were more updates to match the field names of the backend models -- let me know if they should be separated into a different PR.

jcqli commented:
There are linter errors with the bazel files that I'll need to comb through but assigning for a review since it's not blocking anything.

rt4914 commented on data/src/main/java/org/oppia/android/data/backends/gae/model/GaeFeedbackReport.kt:
```diff
@@ -10,8 +10,10 @@ data class GaeFeedbackReport(
   @Json(name = "android_report_info_schema_version") val schemaVersion: Int,
-  /** The date and time in ms that the report was created. */
+  /** The date and time in sec that the report was submitted. */
   @Json(name = "report_submission_timestamp_sec") val reportSubmissionTimestampSec: Int,
```
Should this be `Long` instead of `Int`?

jcqli replied:
This is defined as an int since the backend timestamp for seconds uses an int rather than a double so I kept the consistency. The long => int conversion could happen later, though, (i.e. this could remain a Long and then we perform the conversion in the backend controller) -- would that be preferred? I don't see if there's any benefit to keeping this a Long.

rt4914 reviewed (approved):
Checked all `Gae*` files w.r.t json file and everything looks correct. Approving for code-owner files.

fsharpasharp commented on model/BUILD.bazel:
```diff
@@ -227,6 +228,8 @@ android_library(
     visibility = ["//visibility:public"],
     exports = [
         ":arguments_java_proto_lite",
+        ":profile_java_proto_lite",
```
Don't add these, since we're removing the model library.

jcqli replied:
Done

fsharpasharp commented on model/BUILD.bazel:
```diff
@@ -165,6 +165,7 @@ format_import_proto_library(
 
 java_lite_proto_library(
     name = "feedback_reporting_java_proto_lite",
+    visibility = ["//visibility:public"],
```
Use //:oppia_api_visibility, if possible

jcqli replied:
Done

fsharpasharp commented on utility/src/main/java/org/oppia/android/util/extensions/BUILD.bazel:
```diff
@@ -26,3 +26,15 @@ kt_android_library(
+    ],
+    visibility = ["//:oppia_api_visibility"],
+    deps = [
+        "//model",
```
Don't use the //model dependency. Add the underlying dependencies directly.

jcqli replied:
Done

anandwana001 commented on data/BUILD.bazel:
```diff
@@ -6,6 +6,16 @@ This library provides data to the rest of the application.
+    visibility = ["//:oppia_testing_visibility"],
+    custom_package = "org.oppia.android.data",
+    manifest = "src/test/AndroidManifest.xml",
+)
```
Bazel Lint Fix

```Bazel
android_library(
    name = "data_test_assets",
    testonly = True,
    assets = glob(["src/test/assets/**"]),
    assets_dir = "src/test/assets/",
    custom_package = "org.oppia.android.data",
    manifest = "src/test/AndroidManifest.xml",
    visibility = ["//:oppia_testing_visibility"],
)

```

jcqli replied:
Done

anandwana001 commented on domain/BUILD.bazel:
```diff
@@ -72,12 +72,16 @@ kt_android_library(
     deps = [
         ":dagger",
+        "//data/src/main/java/org/oppia/android/data/backends/gae/api",
+        "//data/src/main/java/org/oppia/android/data/backends/gae/model:model",
```
Bazel lint is failing for this and the Buildifier Auto formatter is formating it like this
`"//data/src/main/java/org/oppia/android/data/backends/gae/model",`

Do we need this `"//data/src/main/java/org/oppia/android/data/backends/gae/model:model",` as it is or we can convert it to `"//data/src/main/java/org/oppia/android/data/backends/gae/model",` ?

If not, we need to check into the Buildifier to prevent such a case from failing.

jcqli replied:
Ah thanks for catching this! Done, changed to `//data/src/main/java/org/oppia/android/data/backends/gae/model`

anandwana001 commented on domain/src/test/java/org/oppia/android/domain/feedbackreporting/FeedbackReportManagementControllerTest.kt:
```diff
@@ -0,0 +1,445 @@
+  }
+
+  @Test
+  fun testController_submitMultipleFeedbackReports_withoutNetwork_savesAllReportsToStore() {
```
Should we add 1 more test for `submitMultipleFeedbackReports_withNetwork`?

jcqli replied:
Ack -- I initially left this case out since it's the same as `testController_submitFeedbackReport_withNetwork_doesNotSaveReportToStore` multiple times but this might catch something.
Test added.

anandwana001 commented on domain/src/test/java/org/oppia/android/domain/feedbackreporting/FeedbackReportManagementControllerTest.kt:
```diff
@@ -0,0 +1,445 @@
+        .setApplicationInfo(applicationInfo)
+        .build()
+    packageInfo.versionName = "1.0"
+    packageInfo.versionCode = 1
```
Should we try `packageInfo.longVersionCode = 1` as `versionCode` is deprecated. 
Check `PackageInfoCompat` class.

jcqli replied:
Done, updated this to be a `Long`. Fixed the proto and extension class as well.

anandwana001 reviewed (changes requested):
looked at the failed static check and test file. Left few thoughts.

jcqli committed:
Address review comments

jcqli committed:
Merge remote-tracking branch 'origin/develop' into create-domain-controller-for-feedback-reporting

jcqli committed:
Fix review comments

jcqli commented:
I'm no longer able to run tests on Gradle -- is anyone else able to repro this? It looks to build fine on both Gradle and Bazel, but I can't run the tests in Gradle. Cleaning and rebuilding didn't help.
![Screen Shot 2021-05-17 at 9 14 31 AM](https://user-images.githubusercontent.com/40325757/118521914-53d7f400-b6f0-11eb-95ed-5e0c029d6e69.png)

jcqli committed:
Enueue two server responses for multiple reports sent

rt4914 commented:
@jcqli I checked on latest develop and I was able to run the app layer and domain layer tests locally on gradle.
<img width="723" alt="Screenshot 2021-05-18 at 5 19 27 PM" src="https://user-images.githubusercontent.com/9396084/118646076-58d69980-b7fd-11eb-81b2-8cb8fa261f23.png">

rt4914 commented:
@jcqli I rechecked for robolectric tests and it was failing. It is fixed in latest develop now. Reference: #3196

jcqli committed:
Merge remote-tracking branch 'origin/develop' into create-domain-controller-for-feedback-reporting

anandwana001 commented on domain/src/test/java/org/oppia/android/domain/feedbackreporting/FeedbackReportManagementControllerTest.kt:
```diff
@@ -0,0 +1,466 @@
+  fun testController_submitFeedbackReport_withNetwork_doesNotSaveReportToStore() {
+    // Enqueue a responses so that the MockWebServer knows when the request is successfully sent.
+    mockWebServer.enqueue(MockResponse().setBody("{}"))
+    networkConnectionUtil.setCurrentConnectionStatus(LOCAL)
```
Should we add test cases for `CELLULAR` type connections?

anandwana001 commented on domain/src/test/java/org/oppia/android/domain/feedbackreporting/FeedbackReportManagementControllerTest.kt:
```diff
@@ -0,0 +1,466 @@
+  fun testController_submitFeedbackReport_withNetwork_doesNotSaveReportToStore() {
+    // Enqueue a responses so that the MockWebServer knows when the request is successfully sent.
+    mockWebServer.enqueue(MockResponse().setBody("{}"))
+    networkConnectionUtil.setCurrentConnectionStatus(LOCAL)
```
Should we add test cases for `CELLULAR` type connections?

jcqli replied:
Done

jcqli committed:
Add test cases for cellular network

fsharpasharp commented on model/BUILD.bazel:
```diff
@@ -227,6 +228,7 @@ android_library(
     visibility = ["//visibility:public"],
     exports = [
         ":arguments_java_proto_lite",
+        ":question_java_proto_lite",
```
Remove this line as well.

jcqli replied:
Done

anandwana001 reviewed (approved):
Test files LGTM

jcqli committed:
Remove question protos from model bazel target

fsharpasharp reviewed (approved):
BUILD.bazel files look good to me.

jcqli committed:
Merge remote-tracking branch 'origin/develop' into create-domain-controller-for-feedback-reporting

jcqli committed:
Merge branch 'develop' into create-domain-controller-for-feedback-reporting

jcqli committed:
Add platform field to report

jcqli committed:
Merge remote-tracking branch 'origin/develop' into create-domain-controller-for-feedback-reporting

jcqli committed:
Exempt language code extensions from tests

jcqli committed:
Specify platform_type in data model

jcqli committed:
Merge remote-tracking branch 'origin/develop' into create-domain-controller-for-feedback-reporting

jcqli committed:
Update service URL to match backend

jcqli committed:
Merge remote-tracking branch 'origin/develop' into create-domain-controller-for-feedback-reporting

BenHenning commented on utility/src/main/java/org/oppia/android/util/extensions/ContextExtensions.kt:
```diff
@@ -10,8 +10,13 @@ fun Context.getVersionName(): String {
 /** Returns the current app version code. */
-fun Context.getVersionCode(): Int {
-  return this.packageManager.getPackageInfo(this.packageName, /* flags= */ 0).versionCode
+fun Context.getLongVersionCode(): Long {
```
Do we really need to use a long? The version code is fixed for a particular binary, and I suspect it's very unlikely for us to exceed 2 billion builds of the app anytime soon (especially considering we don't auto-increment the version code yet).

Though, if the new API only provides a long then maybe just call this 'getVersionCode' since the type distinction isn't that important outside the package manager.

jcqli replied:
`versionCode` is deprecated so we only have the long version (I presume we don't want to use old APIs) -- changed the function name.

BenHenning commented on data/src/main/java/org/oppia/android/data/backends/gae/api/FeedbackReportingService.kt:
```diff
@@ -7,7 +7,7 @@ import retrofit2.http.POST
 /** Service that posts feedback reports to remote storage. */
 interface FeedbackReportingService {
-  @POST("android_feedback_reporting/incoming_report")
+  @POST("appfeedbackreporthandler/incoming_android_report")
```
Just to confirm: are this + the model objects up-to-date with the latest version of the backend handler/were there any changes to that before that haven't yet been reflected here?

jcqli replied:
Yep ([backend PR](https://github.com/oppia/oppia/pull/12367/files#diff-f264a7b7fd28a5d22210dd5ea1d533cc79f2d650db099daeffa4608f393d4e9cR55))

BenHenning commented on data/src/test/java/org/oppia/android/data/backends/gae/RemoteAuthNetworkInterceptorTest.kt:
```diff
@@ -98,6 +98,7 @@ class RemoteAuthNetworkInterceptorTest {
 
   @Test
   fun testNetworkInterceptor_withIncorrectHeaders_setsCorrectHeaders() {
+    mockWebServer.enqueue(MockResponse().setBody("{}"))
```
For context, why this change?

jcqli replied:
For parity with the other test case and reading clarity -- it's setting up the test / networking environment at the beginning rather than in the middle.

BenHenning commented on domain/src/main/java/org/oppia/android/domain/feedbackreporting/FeedbackReportManagementController.kt:
```diff
@@ -0,0 +1,346 @@
+
+/** Controller for uploading feedback reports to remote storage or saving them on disk. */
+@Singleton
+class FeedbackReportManagementController @Inject constructor(
```
For this & its test, could you define a separate kt_android_library & test target at the package level for each? For the most part, we're trying to move toward package-specific build packages now for modules outside of app/ to avoid continuing to move the modularization target for the Bazel migration.

BenHenning commented on domain/src/main/java/org/oppia/android/domain/feedbackreporting/FeedbackReportManagementController.kt:
```diff
@@ -0,0 +1,346 @@
+   *
+   * @return a list of feedback reports to upload to remote storage
+   */
+  suspend fun getCachedReportsList(): MutableList<FeedbackReport> {
```
We probably shouldn't return something that's mutable, especially in a suspend context since it can make data races pretty hard to avoid. Can we just return a standard List here, instead?

jcqli replied:
Done

BenHenning commented on domain/src/main/java/org/oppia/android/domain/feedbackreporting/FeedbackReportManagementController.kt:
```diff
@@ -0,0 +1,346 @@
+      ReportTypeCase.SUGGESTION -> {
+        val suggestion = userSuppliedFeedback.suggestion
+        return GaeUserSuppliedFeedback(
+          reportType = reportType.name.toLowerCase(),
+          category = suggestion.suggestionCategory.name.toLowerCase(),
```
We ought to not convert enums to strings like this. I actually suggest hardcoding them or using constants to ensure the exact correct string that matches the backend is used (otherwise, renaming the enum will unexpectedly break this flow).

Also, we should pass in a Locale for toLowerCase(). Generally, it should be the user's locale if the string is user-supplied, or otherwise the appropriate locale for the string if it's not user-supplied.

Ditto elsewhere (for both).

jcqli replied:
Done, using constants.

Re: locale, would it just be US English? To be readable for maintainers, this wouldn't be user-supplied but it is being determined from the device or app.

BenHenning commented on domain/src/main/java/org/oppia/android/domain/feedbackreporting/FeedbackReportManagementController.kt:
```diff
@@ -0,0 +1,346 @@
+        )
+      }
+      else -> throw IllegalArgumentException(
+        "Encountered unexpected feedback report type: ${userSuppliedFeedback.reportTypeCase.name}"
```
```suggestion
        "Encountered unexpected feedback report type: ${userSuppliedFeedback.reportTypeCase}"
```

Optional: the enum-to-string conversion should keep this error message the way you intend it to be.

Ditto elsewhere.

jcqli replied:
Done

BenHenning commented on domain/src/main/java/org/oppia/android/domain/feedbackreporting/FeedbackReportManagementController.kt:
```diff
@@ -0,0 +1,346 @@
+  ): GaeFeedbackReportingSystemContext {
+    return GaeFeedbackReportingSystemContext(
+      packageVersionName = systemContext.packageVersionName,
+      packageVersionCode = systemContext.packageVersionCode,
```
This one doesn't use the newly updated longVersionCode method?

jcqli replied:
This is getting the version code from a proto, we would use the extension function to create the proto when the report is originally created.

BenHenning commented on domain/src/main/java/org/oppia/android/domain/feedbackreporting/FeedbackReportManagementController.kt:
```diff
@@ -0,0 +1,346 @@
+        topicId = revisionCard.topicId
+        subtopicId = revisionCard.subtopicId
+      }
+      EntryPointCase.NAVIGATION_DRAWER -> {
+        // If entry point is not an exploration player or revision card, leave story values as null.
+      }
+      EntryPointCase.CRASH_DIALOG -> {
+        // If entry point is not an exploration player or revision card, leave story values as null.
+      }
```
```suggestion
      EntryPointCase.NAVIGATION_DRAWER, EntryPointCase.CRASH_DIALOG -> {
        // If entry point is not an exploration player or revision card, leave story values as null.
      }
```

Can be simplified.

jcqli replied:
Done

BenHenning commented on domain/src/main/java/org/oppia/android/domain/feedbackreporting/FeedbackReportManagementController.kt:
```diff
@@ -0,0 +1,346 @@
+
+  // Helper function to retrieve the logcat logs from the device.
+  private fun getLogcatLogs(): List<String> {
+    val logcatReader = consoleLogger.getLogReader()
```
We should make sure that this gets closed.

That being said, you can probably simplify this entire method & ensure the reader gets closed:

```kotlin
return consoleLogger.getLogReader().readLines()
```

I suspect the IOException isn't needed. If it comes up that the file doesn't exist, consider changing the API for ConsoleLogger to return the list of lines & return an empty list if the file doesn't exist (which can be checked on the File object directly rather than trying to open & read a BufferedReader).

jcqli replied:
Done

BenHenning commented on domain/src/main/java/org/oppia/android/domain/feedbackreporting/FeedbackReportManagementController.kt:
```diff
@@ -0,0 +1,346 @@
+    it.getOrDefault(OppiaEventLogs.getDefaultInstance())
+  }
+
+  private val eventsLogList = Transformations.map(eventsLogListLiveData) { oppiaEventLogs ->
+    oppiaEventLogs.eventLogList.map { it.toString() }
+  }.value
```
Does this actually work? I wouldn't expect it to since DataProvider-converted-LiveData should never provide a value unless it has an observer.

In general, we never want to do this. LiveData should only be used by the frontend to safely observe async data channels. Instead, we should return a Deferred or DataProvider from AnalyticsController so that it can be observed or transformed in this class. I suspect that, since you want the value of the operation you'll probably need to use a suspend function to retrieve the result directly from the DataProvider. To do that, you'll need to inject a background dispatcher and use it to launch a new async task to provide you with a suspend context.

BenHenning commented on scripts/assets/test_file_exemptions.textproto:
```diff
@@ -617,6 +617,7 @@ exempted_file_path: "utility/src/main/java/org/oppia/android/util/data/DataProvi
 exempted_file_path: "utility/src/main/java/org/oppia/android/util/data/DataProvidersInjector.kt"
 exempted_file_path: "utility/src/main/java/org/oppia/android/util/data/DataProvidersInjectorProvider.kt"
 exempted_file_path: "utility/src/main/java/org/oppia/android/util/extensions/ContextExtensions.kt"
+exempted_file_path: "utility/src/main/java/org/oppia/android/util/extensions/LanguageCodeExtensions.kt"
```
Could we add tests for this file instead of exempting it?

BenHenning commented on utility/src/main/java/org/oppia/android/util/extensions/LanguageCodeExtensions.kt:
```diff
@@ -0,0 +1,60 @@
+package org.oppia.android.util.extensions
```
This file is introducing more duplication against the enum definitions in profile.proto & the mapping defined in OptionControlsViewModel.kt. If we're going to define one mapping, can we do it in one place (ideally here)?

BenHenning commented on utility/src/main/java/org/oppia/android/util/extensions/LanguageCodeExtensions.kt:
```diff
@@ -0,0 +1,60 @@
+// Extension functions for language codes that convert between the different types of language codes
+// and their String representations.
+
+enum class LanguageCode {
```
Please add KDocs for all public classes/functions/members in this file.

BenHenning commented on domain/src/test/java/org/oppia/android/domain/feedbackreporting/FeedbackReportManagementControllerTest.kt:
```diff
@@ -0,0 +1,519 @@
+import javax.inject.Singleton
+
+// Timestamp in sec for 3/2/21, 12:38pm GMT.
+private const val EARLIER_TIMESTAMP = 1614688684
```
Nit: for this & next one, please add unit to the field names.

BenHenning commented on domain/src/test/java/org/oppia/android/domain/feedbackreporting/FeedbackReportManagementControllerTest.kt:
```diff
@@ -0,0 +1,519 @@
+
+  @Before
+  fun setUp() {
+    networkConnectionUtil = NetworkConnectionUtil(ApplicationProvider.getApplicationContext())
```
This should probably be injected, instead.

BenHenning commented on domain/src/test/java/org/oppia/android/domain/feedbackreporting/FeedbackReportManagementControllerTest.kt:
```diff
@@ -0,0 +1,519 @@
+    setUpTestApplicationComponent()
+    setUpApplicationForContext()
+    setUpFakeLogcatFile()
+    MockitoAnnotations.initMocks(this)
```
Can we use the MockitoJUnit rule instead of manually initing mocks?

BenHenning commented on domain/src/test/java/org/oppia/android/domain/feedbackreporting/FeedbackReportManagementControllerTest.kt:
```diff
@@ -0,0 +1,519 @@
+      timeout = testCoroutineDispatcher.DEFAULT_TIMEOUT_SECONDS,
+      unit = testCoroutineDispatcher.DEFAULT_TIMEOUT_UNIT
+    )
+    request?.let {
```
Can we make this a bit stricter by asserting that the request is non-null? Otherwise, if the test goes wrong & this returns null then I think it will still pass (incorrectly).

Ditto for other tests.

BenHenning commented on domain/src/test/java/org/oppia/android/domain/feedbackreporting/FeedbackReportManagementControllerTest.kt:
```diff
@@ -0,0 +1,519 @@
+    verify(mockReportsStoreObserver, atLeastOnce()).onChanged(reportStoreResultCaptor.capture())
+
+    val reportsList = reportStoreResultCaptor.value.getOrThrow().reportsList
+    assertThat(reportsList.size).isEqualTo(2)
```
```suggestion
    assertThat(reportsList).hasSize(2)
```

Bit more idiomatic.

BenHenning commented on domain/src/test/java/org/oppia/android/domain/feedbackreporting/FeedbackReportManagementControllerTest.kt:
```diff
@@ -0,0 +1,519 @@
+  }
+
+  @Test
+  fun testController_removeCachedReport_noLongerInStore() {
```
Other potential test scenario: calling removeFirstCachedReport() without anything being present.

BenHenning commented on domain/src/main/java/org/oppia/android/domain/feedbackreporting/FeedbackReportManagementController.kt:
```diff
@@ -0,0 +1,346 @@
+   *
+   * @param report a [FeedbackReport] to upload
+   */
+  fun uploadFeedbackReport(report: FeedbackReport) {
```
Should this be private? It seems like we'd always want reports to go through submitFeedbackReport().

If not (particularly because we want to log things through the worker), then we should probably have a few targeted tests that verify this always forces a send regardless of internet connectivity (or put another way: the caller is responsible for calling this in the correct circumstances).

BenHenning commented on domain/src/main/java/org/oppia/android/domain/feedbackreporting/FeedbackReportManagementController.kt:
```diff
@@ -0,0 +1,346 @@
+      return@storeDataAsync feedbackReportDatabase.toBuilder().addReports(report).build()
+    }.invokeOnCompletion {
+      it?.let {
+        consoleLogger.e(
+          FEEDBACK_REPORT_MANAGEMENT_CONTROLLER_TAG,
+          "Failed to store feedback report",
+          it
+        )
```
Are any of these failure cases covered in tests? Consider that the cache file could become corrupted during operations which would result in some of these error cases.

BenHenning commented on domain/src/main/java/org/oppia/android/domain/feedbackreporting/FeedbackReportManagementController.kt:
```diff
@@ -0,0 +1,346 @@
+
+  // Helper function that construct a Moshi data class object for the information provided by the user,
+  // based on the type of Issue report.
+  private fun createGaeUserSuppliedFeedbackForIssue(
```
Just to check: for each of these helper functions, are all of the ``when`` cases covered via tests?

BenHenning commented on domain/src/main/java/org/oppia/android/domain/feedbackreporting/FeedbackReportManagementController.kt:
```diff
@@ -0,0 +1,346 @@
+  }
+
+  // Helper function to retrieve the logcat logs from the device.
+  private fun getLogcatLogs(): List<String> {
```
I suggest testing multiple logcat lines (unless I'm mistaken, I think your current test data only covers a single line which might miss some of the logic being employed here).

BenHenning reviewed (changes requested):
Thanks so much for the PR, and your patience in my very delayed review @jcqli! I've taken a full pass and left some comments--PTAL and let me know if you have any questions or would like clarity on any of my suggestions.

jcqli committed:
Make review changes

jcqli committed:
Create own target for feedback reporting and tests

jcqli committed:
Merge remote-tracking branch 'origin/develop' into create-domain-controller-for-feedback-reporting

jcqli committed:
Nit change, push to update PR

jcqli committed:
Merge remote-tracking branch 'origin/develop' into create-domain-controller-for-feedback-reporting

jcqli committed:
Merge remote-tracking branch 'origin/develop' into create-domain-controller-for-feedback-reporting

jcqli committed:
Update test manifest target, remove unused loaded symbols

jcqli committed:
Merge remote-tracking branch 'origin/develop' into create-domain-controller-for-feedback-reporting

jcqli committed:
Merge remote-tracking branch 'origin/develop' into create-domain-controller-for-feedback-reporting

jcqli committed:
Address some review comments

BenHenning commented:
@jcqli do you have an idea on when you might be able to revive this PR?

jcqli commented:
@BenHenning I'll get some changes and reviewed again next week! Thanks for the patience.

BenHenning commented:
No problem, and sounds good. Thanks!

On Thu, Oct 7, 2021 at 3:42 PM jcqli ***@***.***> wrote:

> Reopened #2878 <https://github.com/oppia/oppia-android/pull/2878>.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/oppia/oppia-android/pull/2878#event-5431407705>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ADDB3PU3WAKVONV2N3R2BNLUFYO3XANCNFSM4Y6YPTEQ>
> .
>
